### PR TITLE
[EAGLE-843] Refactor application shared service registry framework

### PR DESCRIPTION
**Pull request by haoch**
Refactor application shared service registry framework
* Add `Optional<List<Service>> getSharedServices(Config envConfig)` in `ApplicationProvider`
* Move `MRHistoryJobDailyReporter` registry to `MRHistoryJobApplicationProvider` from `ServerApplication`
* Register `getSharedServices` from `ApplicationProvider` to `Environment` in `ServerApplication`
* Add `EmbeddedMailService`

https://issues.apache.org/jira/browse/EAGLE-843

### CODE REVIEW

File: eagle-core/eagle-app/eagle-app-base/src/main/java/org/apache/eagle/app/spi/ApplicationProvider.java [hunk 1]
```diff
@@ -72,7 +75,20 @@
...
+    default Optional<HealthCheck> getManagedHealthCheck(Config config) {
+        return Optional.empty();
+    }
+
+    /**
+     *
+     * @param envConfig server environment config.
+     * @return Server-level shared services.
+     */
+    default Optional<List<Service>> getSharedServices(Config envConfig) {
```
Comment (haoch) [inline eagle-core/eagle-app/eagle-app-base/src/main/java/org/apache/eagle/app/spi/ApplicationProvider.java]: Change#1: add `Optional<List<Service>> getSharedServices(Config envConfig) ` in `ApplicationProvider`

@qingwen220 @wujinhu

**Comment by wujinhu** [comment]
LGTM

File: eagle-jpm/eagle-jpm-mr-history/src/main/java/org/apache/eagle/jpm/mr/history/MRHistoryJobApplicationProvider.java [hunk 2]
```diff
@@ -35,7 +40,16 @@ public MRHistoryJobApplication getApplication() {
     }
 
     @Override
-    public Optional<HealthCheck> getAppHealthCheck(Config config) {
+    public Optional<HealthCheck> getManagedHealthCheck(Config config) {
         return Optional.of(new MRHistoryJobApplicationHealthCheck(config));
     }
+
+    @Override
+    public Optional<List<Service>> getSharedServices(Config envConfig) {
```
Comment (haoch) [inline eagle-jpm/eagle-jpm-mr-history/src/main/java/org/apache/eagle/jpm/mr/history/MRHistoryJobApplicationProvider.java]: Change#2: move `MRHistoryJobDailyReporter ` registry to `MRHistoryJobApplicationProvider` from `ServerApplication` 

@qingwen220 @wujinhu

File: eagle-server/src/main/java/org/apache/eagle/server/ServerApplication.java [hunk 3]
```diff
@@ -107,18 +111,31 @@ public void run(ServerConfig configuration, Environment environment) throws Exce
...
-        environment.lifecycle().manage(appHealthCheckTask);
-
-        if (config.hasPath(MRHistoryJobDailyReporter.SERVICE_PATH)) {
-            Managed jobReportTask = new ApplicationTask(mrHistoryJobDailyReporter);
-            environment.lifecycle().manage(jobReportTask);
+        environment.lifecycle().manage(new ManagedService(applicationHealthCheckService));
+
+        // Load application shared extension services.
+        LOG.debug("Registering application shared extension services");
+        for (ApplicationProvider<?> applicationProvider : applicationProviderService.getProviders()) {
```
Comment (haoch) [inline eagle-server/src/main/java/org/apache/eagle/server/ServerApplication.java]: Change#3, Register `getSharedServices` from `ApplicationProvider` to `Environment` in `ServerApplication`

**Pull request closed without merging**